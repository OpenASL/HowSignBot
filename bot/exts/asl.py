import logging
from urllib.parse import quote_plus

import discord
from discord.ext.commands import Cog, Bot, Context, command, errors

import handshapes
from bot import settings
from bot.utils import get_spoiler_text, did_you_mean

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX

SIGN_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for {word_uppercased}]({signingsavvy})
[🌐 **Spread The Sign** - {word_uppercased}]({spread_the_sign})
[📹 **YouGlish** - ASL videos with {word_uppercased}]({youglish})
Share: {howsign}
"""

SIGN_SPOILER_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for ||{word_uppercased}||]({signingsavvy})
[🌐 **Spread The Sign** - ||{word_uppercased}||]({spread_the_sign})
[📹 **YouGlish** - ASL videos with ||{word_uppercased}||]({youglish})
Share: ||{howsign}||
"""

SIGN_HELP = """Look up a word or phrase

If the word or phrase is sent in spoiler text, i.e. enclosed in `||`, the word will also be blacked out in the reply.
To search multiple words/phrases, separate the values with a comma.

Examples:
{COMMAND_PREFIX}sign tiger
{COMMAND_PREFIX}sign ||tiger||
{COMMAND_PREFIX}sign what's up
{COMMAND_PREFIX}sign church, chocolate, computer
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def word_display(word: str, *, template: str = SIGN_TEMPLATE, max_length: int = 100):
    if len(word) > max_length:
        raise errors.BadArgument("⚠️ Input too long. Try a shorter query.")
    quoted_word = quote_plus(word).lower()
    return template.format(
        word_uppercased=word.upper(),
        howsign=f"https://howsign.app/?s={quoted_word}",
        lifeprint=f"https://www.google.com/search?&q=site%3Alifeprint.com+{quoted_word}",
        handspeak=f"https://www.google.com/search?&q=site%3Ahandspeak.com+{quoted_word}",
        signingsavvy=f"https://www.signingsavvy.com/search/{quoted_word}",
        spread_the_sign=f"https://www.spreadthesign.com/en.us/search/?q={quoted_word}",
        youglish=f"https://youglish.com/pronounce/{quoted_word}/signlanguage/asl",
    )


def sign_impl(word: str):
    logger.info(f"sending links for: '{word}'")
    spoiler = get_spoiler_text(word)
    word = spoiler if spoiler else word
    has_multiple = "," in word
    template = SIGN_SPOILER_TEMPLATE if spoiler else SIGN_TEMPLATE
    if has_multiple:
        words = word.split(",")
        embed = discord.Embed()
        for word in words:
            word = word.strip()
            title = f"||{word.upper()}||" if spoiler else word.upper()
            embed.add_field(name=title, value=word_display(word, template=template))
    else:
        title = f"||{word.upper()}||" if spoiler else word.upper()
        embed = discord.Embed(
            title=title,
            description=word_display(word, template=template),
        )

    return {"embed": embed}


HANDSHAPE_HELP = """Show a random or specific handshape

Examples:
{COMMAND_PREFIX}handshape
{COMMAND_PREFIX}handshape claw5

Enter {COMMAND_PREFIX}handshapes to show a list of handshapes.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshape_impl(name: str):
    logger.info(f"handshape: '{name}'")
    try:
        if name == "random":
            handshape = handshapes.get_random_handshape()
        else:
            handshape = handshapes.get_handshape(name)
    except handshapes.HandshapeNotFoundError:
        logger.info(f"handshape '{name}' not found")
        suggestion = did_you_mean(name, tuple(handshapes.HANDSHAPES.keys()))
        if suggestion:
            return {
                "content": f'"{name}" not found. Did you mean "{suggestion}"? Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }
        else:
            return {
                "content": f'"{name}" not found. Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }

    filename = f"{handshape.name}.png"
    file_ = discord.File(handshape.path, filename=filename)
    embed = discord.Embed(title=handshape.name)
    embed.set_image(url=f"attachment://{filename}")
    return {
        "file": file_,
        "embed": embed,
    }


HANDSHAPES_TEMPLATE = """{handshapes}

Enter `{COMMAND_PREFIX}handshape` to display a random handshape or `{COMMAND_PREFIX}handshape [name]` to display a specific handshape.
"""

HANDSHAPES_HELP = """List handshapes

Enter {COMMAND_PREFIX}handshape to display a random handshape or {COMMAND_PREFIX}handshape [name] to display a specific handshape.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshapes_impl():
    return {
        "content": HANDSHAPES_TEMPLATE.format(
            handshapes=", ".join(handshapes.HANDSHAPES.keys()),
            COMMAND_PREFIX=COMMAND_PREFIX,
        )
    }


class ASL(Cog):
    def __init__(self, bot: Bot):
        self.bot = bot

    @command(name="sign", aliases=("howsign", COMMAND_PREFIX), help=SIGN_HELP)
    async def sign_command(self, ctx: Context, *, word: str):
        await ctx.send(**sign_impl(word))

    @sign_command.error
    async def sign_error(self, ctx: Context, error: Exception):
        # Ignore "??"
        if isinstance(error, errors.MissingRequiredArgument):
            if ctx.invoked_with == COMMAND_PREFIX:
                logger.info(
                    f"no argument passed to {COMMAND_PREFIX}{ctx.invoked_with}. ignoring..."
                )
            else:
                await ctx.send(
                    f"⚠️ Enter a word or phrase to search for after `{COMMAND_PREFIX}{ctx.invoked_with}`."
                )

    @command(name="handshape", aliases=("shape",), help=HANDSHAPE_HELP)
    async def handshape_command(self, ctx: Context, name="random"):
        await ctx.send(**handshape_impl(name))

    @command(name="handshapes", aliases=("shapes",), help=HANDSHAPES_HELP)
    async def handshapes_command(self, ctx: Context):
        logger.info("sending handshapes list")
        await ctx.send(**handshapes_impl())


def setup(bot: Bot) -> None:
    bot.add_cog(ASL(bot))
