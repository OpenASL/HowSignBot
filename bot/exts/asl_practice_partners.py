import asyncio
import datetime as dt
import logging
from contextlib import suppress
from typing import Mapping
from typing import Sequence

import discord
from discord import Color
from discord import Embed
from discord import Guild
from discord import Member
from discord import Message
from discord import VoiceState
from discord.channel import TextChannel
from discord.ext import commands
from discord.ext.commands import Bot
from discord.ext.commands import Cog
from discord.ext.commands import Context
from discord.ext.commands import group
from discord.ext.commands import is_owner

from bot import settings
from bot.database import store
from bot.utils import did_you_mean
from bot.utils.datetimes import EASTERN
from bot.utils.datetimes import utcnow
from bot.utils.gsheets import get_gsheet_client

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX
KICK_MESSAGE = """You've been removed from ASL Practice Partners server due to inactivity from your account.
Don't worry, you can re-join (and we'd love to have you back). You can find the invite link here:
<https://aslpractice.partners>
If you decide to re-join, make sure to post an intro so you don't get kicked again.
"""
UNMUTE_WARNING = (
    "‚ö†Ô∏è You're unmuted in a practice room VC. To maximize inclusivity and learning for all members, "
    "we encourage you to keep your voice off during practice. "
    "ü§ê You can use the text channels to type responses when needed."
)
DAILY_MESSAGE_TIME = dt.time(8, 0)  # Eastern time
DAILY_MEMBER_KICK_TIME = dt.time(12, 0)  # Eastern time
PRUNE_DAYS = 7


def get_next_task_execution_datetime(time_in_eastern: dt.time) -> dt.datetime:
    """Get next execution time for a daily task.

    Returns an eastern-localized datetime.
    """
    now_eastern = dt.datetime.now(EASTERN)
    date = now_eastern.date()
    if now_eastern.time() > time_in_eastern:
        date = now_eastern.date() + dt.timedelta(days=1)
    return EASTERN.localize(dt.datetime.combine(date, time_in_eastern))


def get_gsheet():
    client = get_gsheet_client()
    return client.open_by_key(settings.ASLPP_SHEET_KEY)


def get_sheet_content(worksheet_name: str) -> list[str]:
    sheet = get_gsheet()
    worksheet = sheet.worksheet(worksheet_name)
    # Get all content from first column
    return worksheet.col_values(1)


MAX_NO_INTRO_USERS_TO_DISPLAY = 30


async def make_no_intros_embed():
    members_without_intro = await store.get_aslpp_members_without_intro(
        since=dt.timedelta(days=settings.ASLPP_INACTIVE_DAYS + 1)
    )

    if len(members_without_intro):
        description = "Here are the oldest memberships:\n"
        description += "\n".join(
            tuple(
                f"<@!{member['user_id']}> - Member for {(utcnow() - member['joined_at']).days} days"
                for member in members_without_intro[:MAX_NO_INTRO_USERS_TO_DISPLAY]
            )
        )
    else:
        description = "‚ú® _No members to review_"

    embed = Embed(
        title=f"{len(members_without_intro)} members joined > {settings.ASLPP_INACTIVE_DAYS} days ago, acknowledged the rules, and have not posted an intro",
        description=description,
        color=Color.orange(),
    )
    embed.set_footer(
        text=f"These members will automatically be kicked at noon Eastern time. Use {COMMAND_PREFIX}aslpp active <members> to prevent members from getting kicked.\n"
        f"Members who have not logged on in {PRUNE_DAYS} days and have no roles will also be pruned."
    )
    return embed


def get_tags():
    logger.info("fetching tags")
    sheet = get_gsheet()
    worksheet = sheet.worksheet("tags")
    return {
        tag.lower(): {"title": title, "description": content}
        for tags, title, content in worksheet.get_all_values()[1:]  # first row is header
        for tag in tags.split()
    }


def format_role_table(data: Sequence[tuple[str, int]], member_count: int) -> str:
    longest_label_length = max(len(label) for label, _ in data)
    return "\n".join(
        [
            f"{label.rjust(longest_label_length)} ‚ñè {count:#2d} | {round(count / member_count * 100)}%"
            for label, count in data
        ]
    )


def make_role_table(guild: Guild, label: str, role_ids: Sequence[int]) -> str:
    data = []
    for role_id in role_ids:
        role = guild.get_role(role_id)
        if not role:
            continue
        data.append((role.name, len(role.members)))
    table = format_role_table(data, guild.member_count)
    return f"```\n{label.upper()}\n{table}\n```"


class AslPracticePartners(Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.tags: dict[str, dict[str, str]] = {}
        self.unmute_warnings: dict[int, dt.datetime] = {}

    def cog_check(self, ctx: Context):
        if not bool(ctx.guild) or ctx.guild.id != settings.ASLPP_GUILD_ID:
            raise commands.errors.CheckFailure(
                f"‚ö†Ô∏è `{COMMAND_PREFIX}{ctx.invoked_with}` must be run within the ASL Practice Partners server (not a DM)."
            )
        return True

    @group(
        name="tag",
        aliases=("tags", "t"),
        invoke_without_command=True,
        hidden=True,
        help="Display a tag or the list of available tags",
    )
    async def tag_group(self, ctx: Context, *, tag_name: str | None):
        if not tag_name:
            await self.display_tags(ctx)
            return
        tag_name = tag_name.lower()
        if tag_name not in self.tags:
            suggestion: str | None = did_you_mean(tag_name, tuple(self.tags.keys()))
            if not suggestion:
                await ctx.reply(f'‚ö†Ô∏è No tag matching "{tag_name}"')
                return
            else:
                tag_name = suggestion
        await ctx.reply(embed=Embed.from_dict(self.tags[tag_name]))

    @tag_group.command(
        "update",
        aliases=("sync",),
        hidden=True,
        help="Sync the tags with the spreadsheet",
    )
    @commands.has_permissions(kick_members=True)  # Staff
    async def update_tags(self, ctx: Context):
        await ctx.channel.trigger_typing()
        self.tags = get_tags()
        await ctx.reply("‚úÖ Updated tags.")

    async def display_tags(self, ctx: Context):
        embed = Embed(
            title="Tags",
            description="\n".join(sorted(f"**¬ª** {tag}" for tag in self.tags)),
        )
        embed.set_footer(text=f"To show a tag, type {COMMAND_PREFIX}tag <tagname>.")
        await ctx.reply(embed=embed)

    @group(name="aslpp", hidden=True)
    async def aslpp_group(self, ctx: Context):
        pass

    # TODO: Allow mods/admins to use these commands
    @aslpp_group.command(name="faq", hidden=True)
    @is_owner()
    async def faq_command(self, ctx: Context, channel: TextChannel):
        await ctx.channel.trigger_typing()
        for content in get_sheet_content("faq"):
            await channel.send(content)
            # Sleep to ensure messages are always displayed in the correct order
            await asyncio.sleep(1)
        await ctx.reply("üôå FAQ posted")

    @aslpp_group.command(name="rules", hidden=True)
    @is_owner()
    async def rules_command(self, ctx: Context, channel: TextChannel):
        await ctx.channel.trigger_typing()
        for content in get_sheet_content("rules"):
            await channel.send(content)
            # Sleep to ensure messages are always displayed in the correct order
            await asyncio.sleep(1)
        await ctx.reply("üôå Rules posted")

    @aslpp_group.command(name="welcome", hidden=True)
    @is_owner()
    async def welcome_command(self, ctx: Context, channel: TextChannel):
        await ctx.channel.trigger_typing()
        for content in get_sheet_content("welcome"):
            await channel.send(content)
            # Sleep to ensure messages are always displayed in the correct order
            await asyncio.sleep(1)
        await ctx.reply("üôå Welcome message posted")

    @aslpp_group.command(name="video", hidden=True)
    @is_owner()
    async def video_command(self, ctx: Context, channel: TextChannel):
        await ctx.channel.trigger_typing()
        for content in get_sheet_content("video-etiquette"):
            await channel.send(content)
            # Sleep to ensure messages are always displayed in the correct order
            await asyncio.sleep(1)
        await ctx.reply("üôå Video etiquette message posted")

    @aslpp_group.command(name="syncdata", hidden=True)
    @is_owner()
    async def sync_data_command(self, ctx: Context):
        await ctx.channel.trigger_typing()
        channel = self.bot.get_channel(settings.ASLPP_INTRODUCTIONS_CHANNEL_ID)
        await store.clear_aslpp_intros()
        async for message in channel.history(limit=None):
            logger.info(f"storing intro record {message.id}")
            await store.add_aslpp_intro(
                message_id=message.id,
                user_id=message.author.id,
                posted_at=message.created_at,
            )

        role = ctx.guild.get_role(settings.ASLPP_ACKNOWLEDGED_RULES_ROLE_ID)
        await store.clear_aslpp_members()
        for member in role.members:
            if member.bot:
                continue
            logger.info(f"storing member {member.id}")
            await store.add_aslpp_member(user_id=member.id, joined_at=member.joined_at)
        logger.info("finished syncing data")
        await ctx.reply("üôå Synced data")

    @aslpp_group.command(name="nointros", aliases=("nointro",), hidden=True)
    @commands.has_permissions(kick_members=True)
    async def no_intros_command(self, ctx: Context):
        await ctx.channel.trigger_typing()
        embed = await make_no_intros_embed()
        await ctx.send(embed=embed)

    @aslpp_group.command(name="active", hidden=True)
    @commands.has_permissions(kick_members=True)
    async def active_command(self, ctx: Context, members: commands.Greedy[Member]):
        await store.mark_aslpp_members_active(user_ids=[m.id for m in members])
        await ctx.reply(f"Marked {len(members)} member(s) active.")

    @aslpp_group.command(name="inactive", hidden=True)
    @commands.has_permissions(kick_members=True)
    async def inactive_command(self, ctx: Context, members: commands.Greedy[Member]):
        await store.mark_aslpp_members_inactive(user_ids=[m.id for m in members])
        await ctx.reply(f"Marked {len(members)} member(s) inactive.")

    @aslpp_group.command(name="kick", hidden=True)
    @commands.has_permissions(kick_members=True)
    async def kick_command(
        self, ctx: Context, targets: commands.Greedy[TextChannel | Member]
    ):
        num_kicked = 0
        for target in targets:
            with suppress(discord.errors.Forbidden):  # user may not allow DMs from bot
                await target.send(KICK_MESSAGE)
            if isinstance(target, Member):
                logger.info(f"kicking member {target.id}")
                await ctx.guild.kick(target, reason="Inactivity")
                num_kicked += 1

        await ctx.reply(f"Kicked {num_kicked} members.")

    async def _kick_inactive(
        self,
        ctx: Context | TextChannel,
        *,
        members_without_intro: list[Mapping],
    ):
        num_kicked = 0
        guild = ctx.guild
        for member_record in members_without_intro[:MAX_NO_INTRO_USERS_TO_DISPLAY]:
            user_id = member_record["user_id"]
            member = guild.get_member(user_id)
            if not member:
                continue
            with suppress(discord.errors.Forbidden):  # user may not allow DMs from bot
                await member.send(KICK_MESSAGE)
            logger.info(f"kicking member {member.id}")
            await guild.kick(member, reason="Inactivity")
            num_kicked += 1
        logger.info(
            f"pruning members who have not logged on in {PRUNE_DAYS} days and have no roles"
        )
        num_pruned = await guild.prune_members(days=PRUNE_DAYS)
        total_kicked = num_kicked + num_pruned
        return total_kicked

    @aslpp_group.command(name="kickall", hidden=True)
    @commands.has_permissions(kick_members=True)
    async def kickall_command(self, ctx: Context):
        members_without_intro = await store.get_aslpp_members_without_intro(
            since=dt.timedelta(days=settings.ASLPP_INACTIVE_DAYS + 1)
        )
        n_members_to_prune = await ctx.guild.estimate_pruned_members(days=PRUNE_DAYS)
        if not len(members_without_intro) and not n_members_to_prune:
            await ctx.send("‚ú® _No members to kick_")
            return
        num_kicked = await self._kick_inactive(
            ctx, members_without_intro=members_without_intro
        )
        await ctx.reply(f"Kicked {num_kicked} members.")

    @group(
        name="rolestats",
        aliases=("rs",),
        invoke_without_command=True,
        hidden=True,
        doc="Show graphs of role member counts",
    )
    @commands.has_permissions(kick_members=True)  # Staff
    async def role_table_group(self, ctx: Context):
        await ctx.channel.trigger_typing()
        await ctx.send(content=self.make_role_table_skill(ctx.guild))
        await ctx.send(content=self.make_role_table_hearing_spectrum(ctx.guild))
        await ctx.send(content=self.make_role_table_age(ctx.guild))

    def make_role_table_skill(self, guild: Guild):
        return make_role_table(guild, "Skill", settings.ASLPP_SKILL_ROLE_IDS)

    @role_table_group.command("skill", hidden=True)
    @commands.has_permissions(kick_members=True)  # Staff
    async def role_table_skill(self, ctx: Context):
        await ctx.channel.trigger_typing()
        await ctx.reply(content=self.make_role_table_skill(ctx.guild))

    def make_role_table_hearing_spectrum(self, guild: Guild):
        return make_role_table(
            guild, "Hearing Spectrum", settings.ASLPP_HEARING_SPECTRUM_ROLE_IDS
        )

    @role_table_group.command("hearingspectrum", hidden=True)
    @commands.has_permissions(kick_members=True)  # Staff
    async def role_table_hearing_spectrum(self, ctx: Context):
        await ctx.channel.trigger_typing()
        await ctx.reply(content=self.make_role_table_hearing_spectrum(ctx.build))

    def make_role_table_age(self, guild: Guild):
        return make_role_table(guild, "Age", settings.ASLPP_AGE_ROLE_IDS)

    @role_table_group.command("age", hidden=True)
    @commands.has_permissions(kick_members=True)  # Staff
    async def role_table_age(self, ctx: Context):
        await ctx.channel.trigger_typing()
        await ctx.reply(content=self.make_role_table_age(ctx.guild))

    @Cog.listener()
    async def on_message(self, message: Message) -> None:
        if message.guild and message.guild.id != settings.ASLPP_GUILD_ID:
            return
        if message.channel.id == settings.ASLPP_INTRODUCTIONS_CHANNEL_ID:
            if await store.has_aslpp_intro(message.author.id):
                logger.debug(f"{message.author.id} already has intro")
            else:
                logger.info(f"storing intro record {message.id}")
                await store.add_aslpp_intro(
                    message_id=message.id,
                    user_id=message.author.id,
                    posted_at=message.created_at,
                )

    @Cog.listener()
    async def on_member_remove(self, member: Member) -> None:
        if member.guild.id != settings.ASLPP_GUILD_ID:
            return
        logger.info(f"removing data for aslpp member {member.id}")
        await store.remove_aslpp_member(user_id=member.id)

    @Cog.listener()
    async def on_member_update(self, before: Member, after: Member) -> None:
        if after.guild.id != settings.ASLPP_GUILD_ID:
            return
        before_role_ids = {role.id for role in before.roles}
        after_role_ids = {role.id for role in after.roles}
        added_role_ids = after_role_ids - before_role_ids
        removed_role_ids = before_role_ids - after_role_ids
        if settings.ASLPP_ACKNOWLEDGED_RULES_ROLE_ID in added_role_ids:
            logger.info(f"aslpp member acknowledged rules. storing member {after.id}")
            await store.add_aslpp_member(user_id=after.id, joined_at=after.joined_at)
        elif settings.ASLPP_ACKNOWLEDGED_RULES_ROLE_ID in removed_role_ids:
            logger.info(
                f"aslpp member acknowledged rules role removed. removing member {after.id}"
            )
            await store.remove_aslpp_member(user_id=after.id)

    @Cog.listener()
    async def on_voice_state_update(
        self, member: Member, before: VoiceState, after: VoiceState
    ) -> None:
        if member.guild.id != settings.ASLPP_GUILD_ID:
            return
        if not settings.ASLPP_ENABLE_UNMUTE_WARNING:
            return
        if after.channel is None:  # member left VC
            return
        if not after.self_mute:  # member unmuted
            # Send warning iff it hasn't been sent to the member in the past hour
            if member.id in self.unmute_warnings:
                last_sent = self.unmute_warnings[member.id]
                if (utcnow() - last_sent) > dt.timedelta(hours=1):
                    await self._send_unmute_warning(member)
            else:
                await self._send_unmute_warning(member)

    async def _send_unmute_warning(self, member: Member):
        try:
            logger.info(f"sending unmute warning to member {member.id}")
            await member.send(content=UNMUTE_WARNING)
            self.unmute_warnings[member.id] = utcnow()
        except Exception:
            logger.exception(f"could not send unmute warning to member {member.id}")

    @Cog.listener()
    async def on_ready(self):
        self.bot.loop.create_task(self.daily_message())
        self.bot.loop.create_task(self.daily_member_kick())
        self.tags = get_tags()

    async def daily_message(self):
        while True:
            next_execution_time = get_next_task_execution_datetime(DAILY_MESSAGE_TIME)
            logger.info(
                f"aslpp staff daily message will be sent at at {next_execution_time.isoformat()}"
            )
            await discord.utils.sleep_until(
                next_execution_time.astimezone(dt.timezone.utc)
            )
            channel = self.bot.get_channel(settings.ASLPP_BOT_CHANNEL_ID)
            await channel.send(content=self.make_role_table_skill(channel.guild))
            await channel.send(
                content=self.make_role_table_hearing_spectrum(channel.guild)
            )
            await channel.send(content=self.make_role_table_age(channel.guild))
            embed = await make_no_intros_embed()
            await channel.send(embed=embed)
            logger.info("sent aslpp staff daily message")

    async def daily_member_kick(self):
        while True:
            next_execution_time = get_next_task_execution_datetime(DAILY_MEMBER_KICK_TIME)
            logger.info(
                f"aslpp inactive members will be kicked at {next_execution_time.isoformat()}"
            )
            await discord.utils.sleep_until(
                next_execution_time.astimezone(dt.timezone.utc)
            )
            members_without_intro = await store.get_aslpp_members_without_intro(
                since=dt.timedelta(days=settings.ASLPP_INACTIVE_DAYS + 1)
            )
            channel = self.bot.get_channel(settings.ASLPP_BOT_CHANNEL_ID)
            guild: Guild = channel.guild
            n_members_to_prune = await guild.estimate_pruned_members(days=PRUNE_DAYS)
            if not len(members_without_intro) and not bool(n_members_to_prune):
                logger.info("no inactive aslpp members to kick")
                continue
            await channel.send(content="ü•æ _Kicking inactive members_...")
            await self._kick_inactive(
                channel, members_without_intro=members_without_intro
            )
            logger.info("cleared aslpp inactive members")


def setup(bot: Bot) -> None:
    bot.add_cog(AslPracticePartners(bot))
