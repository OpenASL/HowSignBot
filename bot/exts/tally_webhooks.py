from __future__ import annotations

import asyncio
import hashlib
import logging
from contextlib import suppress
from typing import Any
from typing import NamedTuple
from typing import TypedDict

import disnake
from aiohttp import web
from disnake.ext.commands import Bot

from bot import settings
from bot.database import store
from bot.utils.gsheets import get_gsheet_client
from bot.utils.ui import LinkView

logger = logging.getLogger(__name__)

WORKSHEET_NAME = "feedback-submissions"


async def add_submission_to_gsheet(
    *, worksheet, bot: Bot, submission: Submission, discord_user_id: int | None
):
    join_month, role_names = "", ""
    if discord_user_id:
        aslpp_member = await store.get_aslpp_member(discord_user_id)
        if aslpp_member:
            join_month = aslpp_member["joined_at"].strftime("%Y-%m")
        guild: disnake.Guild = bot.get_guild(settings.ASLPP_GUILD_ID)
        member = guild.get_member(discord_user_id)
        if member:
            # Skip the @everyone role
            roles = member.roles[1:]
            role_names = "|".join([role.name for role in roles])
    row = (join_month, role_names) + submission
    worksheet.append_row(row)


def hash_int(integer: int) -> str:
    hash_object = hashlib.sha1(str(integer).encode("utf8"))
    return hash_object.hexdigest()


USER_ID_KEY = "question_wbj5B2_e34181f0-00e1-47ae-a3a8-063f16f5821d"
FIELD_KEYS_TO_SUBMISSION_FIELDS = {
    "question_nWEOGP": "practice_session_participation",
    "question_wa5QbE": "proficiency_improvement",
    "question_nP1R1x": "ways_improved",
    "question_mRMWKP": "does_host",
    "question_woe96M": "wants_to_host",
    "question_nG9evz": "staff_can_follow_up",
    "question_mO4QDA": "discord_username",
    "question_m6j8bO": "practice_suggestion",
    "question_w7LRB9": "server_suggestion",
    "question_3y2PKp": "source",
}


class Submission(NamedTuple):
    # NOTE: The order of fields must match the left-to-right order in the Google Sheet
    created_at: str
    hashed_user_id: str
    # NOTE: integer values get stringified
    practice_session_participation: str
    proficiency_improvement: str
    ways_improved: str
    does_host: str
    wants_to_host: str
    staff_can_follow_up: str
    discord_username: str
    practice_suggestion: str
    server_suggestion: str
    source: str


class Option(TypedDict):
    id: str
    text: str


class Field(TypedDict, total=False):
    key: str
    label: str
    type: str
    value: Any
    options: list[Option]


# NOTE: Not all field types are supported here--just the ones that are actually used
def get_field_value(field: Field) -> str:
    value = field["value"]
    if value is None:
        return ""
    if field["type"] == "MULTIPLE_CHOICE":
        options = field["options"]
        option_map = {option["id"]: option["text"] for option in options}
        return option_map[value] or ""
    else:
        return str(value) if value else ""


async def handle_tally_webhook(bot: Bot, data: dict):
    submission, discord_user_id = make_submission(data)
    if not submission:
        return
    logger.info(f"adding feedback submission to gsheet: {submission}")
    client = get_gsheet_client()
    sheet = client.open_by_key(settings.ASLPP_SHEET_KEY)
    worksheet = sheet.worksheet(WORKSHEET_NAME)
    await add_submission_to_gsheet(
        worksheet=worksheet,
        bot=bot,
        submission=submission,
        discord_user_id=discord_user_id,
    )

    url = f"https://docs.google.com/spreadsheets/d/{settings.ASLPP_SHEET_KEY}/edit#gid={worksheet.id}"
    await bot.get_channel(settings.ASLPP_BOT_CHANNEL_ID).send(
        "ðŸ™Œ A member submitted the survey!",
        view=LinkView("Survey Results", url=url),
    )


def make_submission(
    data: dict[str, Any],
) -> tuple[Submission | None, int | None]:
    if not data.get("eventType") == "FORM_RESPONSE":
        return None, None
    try:
        fields: list[Field] = data["data"]["fields"]
        created_at = data["data"]["createdAt"]
    except KeyError:
        return None, None

    discord_user_id = None
    submission_fields: dict[str, str] = {}
    for field in fields:
        key = field["key"]
        if key == USER_ID_KEY:
            with suppress(ValueError):
                discord_user_id = int(field["value"])
        elif key in FIELD_KEYS_TO_SUBMISSION_FIELDS:
            submission_field_key = FIELD_KEYS_TO_SUBMISSION_FIELDS[key]
            submission_fields[submission_field_key] = get_field_value(field)

    return (
        Submission(
            created_at=created_at,
            hashed_user_id=hash_int(discord_user_id) if discord_user_id else "",
            **submission_fields,
        ),
        discord_user_id,
    )


def setup(bot: Bot) -> None:
    async def tally(request):
        data = await request.json()
        # Run the handler logic asynchronously to respond quickly
        asyncio.create_task(handle_tally_webhook(bot, data))
        return web.Response(body="", status=200)

    bot.app.add_routes([web.post("/tally", tally)])
