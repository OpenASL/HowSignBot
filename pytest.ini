[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
