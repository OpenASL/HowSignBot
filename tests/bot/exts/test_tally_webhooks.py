import json
from pathlib import Path

import pytest

from bot.exts.tally_webhooks import make_submission


@pytest.mark.parametrize(
    "example_filename", ("example_tally_webhook.json", "example_tally_webhook2.json")
)
def test_make_submission(snapshot, example_filename):
    example_webhook_path = Path(__file__).parent / example_filename
    with example_webhook_path.open("r") as fp:
        webhook_data = json.load(fp)
    submission = make_submission(webhook_data)
    assert submission == snapshot
