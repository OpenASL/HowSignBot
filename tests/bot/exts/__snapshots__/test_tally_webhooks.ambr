# name: test_make_submission[example_tally_webhook.json]
  <class 'tuple'> (
    <class 'Submission'> (
      created_at='2021-12-30T21:04:10.000Z',
      discord_username='sloria',
      does_host='No',
      hashed_user_id='6112bec78ab57fb4e1ccbc951d880034d5fd57a3',
      meeting_absence_reasons="I don't know enough ASL to participate|The meetings are too crowded",
      meeting_absence_reasons_other='',
      meeting_participation='4',
      meeting_suggestion='More cowbell',
      other_feedback='Nah',
      proficiency_improvement='5',
      server_like='Friendly community',
      server_suggestion='Lots more cowbell',
      source='Duck Duck Go',
      staff_can_follow_up='Yes',
      wants_to_host='Yes',
      ways_improved='Vocab, grammar, sign variations',
    ),
    112053628230342848,
  )
---
# name: test_make_submission[example_tally_webhook2.json]
  <class 'tuple'> (
    <class 'Submission'> (
      created_at='2021-12-30T21:19:52.000Z',
      discord_username='',
      does_host='Yes',
      hashed_user_id='',
      meeting_absence_reasons="There aren't many people that join when I want to practice",
      meeting_absence_reasons_other='',
      meeting_participation='3',
      meeting_suggestion='',
      other_feedback='',
      proficiency_improvement='1',
      server_like='',
      server_suggestion='',
      source="Don't remember",
      staff_can_follow_up='',
      wants_to_host='',
      ways_improved='',
    ),
    None,
  )
---
